### ISSUE TITLE
NoSuchResource when accessing a protected leaf resource

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4018]
|<img alt="unshift's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| unshift reported|
|-|-|
|Trac ID|trac#4018|
|Type|defect|
|Created|2009-09-12 20:00:51Z|

When protecting a Site object which consists of a root Resource and multiple leaf resources, accessing a leaf node will result in a NoSuchResource exception.

For example, if I guard my root object, and then add to it /foo/die, then accessing /foo/die will try to render /.  Setting a breakpoint on getChildWithDefault shows that self.children has "foo" in it (meaning / is being rendered) -- there's no traversal mechanism there.

This might be the way it's supposed to work (and the unit tests only really test this case, as far as I can tell from a quick look), but the documentation notes that the auth mechanism is supposed to be a drop-in solution, so either the code or doc is mistaken.

Attachments:

* [4018.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/b53/b53209f0e4d2b2335d726e4e381511d59d22a31f/53a1b6e82084f0aa96a6196f7de90e1014e008d3.py) (1509 bytes) - added by unshift on 2009-09-12 20:01:47Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__4018 4018
type__defect defect
reporter__unshift unshift
priority__normal normal
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__invalid invalid
component__web web
keywords__ 
time__1252785651000000 1252785651000000
changetime__1252863022000000 1252863022000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|<img alt="unshift's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|unshift commented|
|-|-|

I think I tracked this one down.

resource.getChildForRequest is called to find the node that should be used to handle the request:

```
def getChildForRequest(resource, request):
    """Traverse resource tree to find who will handle the request."""
    while request.postpath and not resource.isLeaf:
        pathElement = request.postpath.pop(0)
        request.prepath.append(pathElement)
        resource = resource.getChildWithDefault(pathElement, request)
    return resource
```

Eventually resource will be a ResourceWrapper.

ResourceWrapper.getChildWithDefault is a pass-through straight to the wrapped resource's getChildWithDefault method.  With authentication enabled, the wrapped resource is going to be whatever the guard was set on -- in the example above and in the attached code, the guard was set on /, but really getChildWithDefault should be called on the Resource representing /foo.  In that case, the child node ("die") that *should* go back to getChildForRequest won't be there, and a NoSuchResource will be returned.

It seems this is a fundamental issue with the authentication strategy in general, since the notion of wrapping a sub-tree means either breaking the simple abstraction and putting node search code into getChildWithDefault, or changing the way that logins are handled.  Since the HTTPAuthSessionWrapper's login code just gets the avatar from the realm, and the avatar is an entire tree, there's no concise way to get the desired node directly from the avatar.  Not without being fancy on the caller's side, anyway.

**Comment by twisted-trac** [comment]
|<img alt="unshift's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|unshift set status to `closed`|
|-|-|

This shows up in 8.2 but seems fixed as of r27323

**Comment by twisted-trac** [comment]
|<img alt="unshift's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|unshift commented|
|-|-|

It's worth noting, the fix may have come from the fix to [#3679](https://github.com/twisted/twisted/issues/3679)

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
